Add stubs for signing updates and images

gtkramer opened the pull request:
The signing solution required for each user of clr-distro-factory will
vary greatly.  A different solution may be needed for signing updates
than from signing images as well.

If a function is defined as a part of the user's configuration
repository, and this function has as a parameter interface the input and
output files as the first and second parameters, respectively, then this
function is used during the build process.  Users may define a
sign_update function as well as a sign_image function.

This abstracts the particular implementation of the signing solution
away from the build process and allows it to be defined entirely as a
configuration.

Signed-off-by: George T Kramer <george.t.kramer@intel.com>

gtkramer committed:
Add stubs for signing updates and images

The signing solution required for each user of clr-distro-factory will
vary greatly.  A different solution may be needed for signing updates
than from signing images as well.

If a function is defined as a part of the user's configuration
repository, and this function has as a parameter interface the input and
output files as the first and second parameters, respectively, then this
function is used during the build process.  Users may define a
sign_update function as well as a sign_image function.

This abstracts the particular implementation of the signing solution
away from the build process and allows it to be defined entirely as a
configuration.

Signed-off-by: George T Kramer <george.t.kramer@intel.com>

mbelluzzo commented on release/images.sh:
```diff
@@ -37,18 +37,11 @@ checksum_and_sign() {
-            "${IMG_SIGN_CMD}" "${chksum_file}" "${chksum_file}.sig"
-        elif [[ -s "${IMG_SIGN_KEY}" ]]; then
-            log "Signing (openssl)" "${chksum_file}"
-            openssl dgst -sha1 -sign "${IMG_SIGN_KEY}" -out "${chksum_file}.sig" "${chksum_file}"
```
We would still like to support a default signing method for images without the need to provide custom command. 
Any particular reason for removing this support?

gtkramer replied:
I originally thought we could abstract all signing away into a function, since user requirements for signing tend to be pretty specific.  But after speaking offline, I see we want to provide a default signing method rather than leaving it entirely up to the user.  Then if users need to deviate from the default, then they can use the function approach.

mbelluzzo commented on release/images.sh:
```diff
@@ -37,18 +37,11 @@ checksum_and_sign() {
-
-        if [[ -s "${chksum_file}.sig" ]]; then
-            _output+="${chksum_file}.sig " # space is not a mistake
+        if function_exists sign_image; then
```
Keep the logging similar to what was done originally.

mbelluzzo commented on release/mixer.sh:
```diff
@@ -53,18 +53,34 @@ build_bundles() {
     local mix_ver="$1"
 
     section "Build 'Update' Content"
+    local mixer_opts_update=""
```
Maybe start with "--skip-format-check" instead of "" ?

mbelluzzo commented on release/mixer.sh:
```diff
@@ -53,18 +53,34 @@ build_bundles() {
+    fi
+    # shellcheck disable=SC2086
+    sudo_mixer_cmd build update --skip-format-check ${mixer_opts_update}
+    if function_exists sign_update; then
```
Add blank line before the if since the result of previous operation is not being evaluated here. Makes code easier to read.

gtkramer replied:
Hmm, I didn't have a blank line here to group usage of `sudo`, since everything requires it.  I was also thinking that since we are replacing a function normally done as a part of `mixer build update` that these should be together as a unit for producing an update.

gtkramer replied:
Added blank line

mbelluzzo commented on release/mixer.sh:
```diff
@@ -53,18 +53,34 @@ build_bundles() {
-        sudo_mixer_cmd build update --skip-format-check
+        mixer_opts_update+=" --min-version=${mix_ver}"
+    fi
+    if function_exists sign_update; then
```
blank line before if

gtkramer replied:
I added one after the section so all of the option building code is grouped together?

mbelluzzo commented on release/mixer.sh:
```diff
@@ -53,18 +53,34 @@ build_bundles() {
+        pushd "update/www/${mix_ver}" > /dev/null
+        sign_update Manifest.MoM Manifest.MoM.sig
+        sudo rm -f Manifest.MoM.tar
+        sudo tar -cJf Manifest.MoM.tar Manifest.MoM Manifest.MoM.sig
```
all the joy of mixer ☺

@bryteise will love this

gtkramer replied:
The joy of serving swupd as our master ;)

mbelluzzo commented on release/mixer.sh:
```diff
@@ -53,18 +53,34 @@ build_bundles() {
+    if function_exists sign_update; then
+        mixer_opts_bundles+=" --no-signing"
+    fi
+    # shellcheck disable=SC2086
```
blank line before comment line.

mbelluzzo commented on release/prologue.sh:
```diff
@@ -63,7 +63,25 @@ Upstream URL:
 
+echo
+echo '== Signing =='
+echo 'sign_update function exists:'
```
s/exists/provided/ for the logging. Maybe even: "Update Custom Signing provide?" similar to "Is this an upstream?"

gtkramer commented:
Rebased

gtkramer commented:
Made more of the changes we talked about offline.  Namely, I've removed the `SIGNING_DIR` variable, adjusted some of the whitespacing, and added the swupd cert copy in the prologue.  I've also printed out the path to the swupd cert in the prologue, but not sure if that's needed.

mbelluzzo commented on release/prologue.sh:
```diff
@@ -27,11 +27,15 @@ LOG_INDENT=1 fetch_config_repo
 mkdir -p "${MIXER_DIR}"
 rm -rf "${MIXER_DIR}/local-yum"
 mkdir -p "${MIXER_DIR}/local-yum"
+if function_exists sign_update; then
```
I think is better to check if SWUPD_CERT is not Null/Undefined than to rely on the function implementation.

mbelluzzo commented on release/prologue.sh:
```diff
@@ -27,11 +27,15 @@ LOG_INDENT=1 fetch_config_repo
 rm -rf "${MIXER_DIR}/local-yum"
 mkdir -p "${MIXER_DIR}/local-yum"
+if function_exists sign_update; then
+    cp "${SWUPD_CERT}" "${MIXER_DIR}/Swupd_Root.pem"
```
actually, why not doing it inside if of line 75?

mbelluzzo replied:
and use :? expansion

gtkramer commented:
I just did a final round of testing and rebased.  Things looks good from my end.

mbelluzzo commented:
LGTM

mbelluzzo commented:
Merged, thanks!
